### PR TITLE
IdentityX+Omeda customer data resync

### PR DESCRIPTION
**Pull request by solocommand**
Currently, the `onActiveContext` hook is firing after the response is sent (due to streaming), meaning that the cookie changes that are being attempted are never actually being written. The first set of changes here moves this logic to a middleware to ensure that it can be attempted before the streamed response is sent. Since this hook now has no logic, it was removed entirely.

A new middleware was added to handle resyncing of customer data: setting custom boolean and select field answers based on the external ids present in the IdentityX user's related Omeda customer record. This works by setting an `__idx_omeda_resync` cookie with a 1 week expiration. This expiration can be customized via the IdentityX configuration file, using the key `userResyncInterval`:
```js
module.exports = new IdentityXConfiguration({
  // ...
  userResyncInterval: 24 * 60 * 60 * 1000, // Run daily
});
```
Finally, a significant portion of the logic used in `onLoginLinkSent` was extracted so it can be re-used by the resync middleware.

### CODE REVIEW

**Comment by solocommand** [comment]
[Add configurable external provider sync hook/middleware handler](https://trello.com/c/9LrAEgh5/17-add-configurable-external-provider-sync-hook-middleware-handler)

**Review by zarathustra323 — COMMENTED** [review]
Can we table this until I'm back from Portland on Tuesday? Would like to go over it in person.

File: packages/marko-web-omeda-identity-x/middleware/set-olytics-cookie.js [hunk 1]
```diff
@@ -1,14 +1,11 @@
...
 const findEncryptedId = require('../external-id/find-encrypted-customer-id');
 
 module.exports = ({
   brandKey,
-  activeContext,
-  req,
-  res,
-}) => {
-  const user = get(activeContext, 'user');
+}) => async (req, res, next) => {
```
Comment (zarathustra323) [inline packages/marko-web-omeda-identity-x/middleware/set-olytics-cookie.js]: Wrap this in the `asyncRoute` helper

File: packages/marko-web-omeda-identity-x/middleware/resync-customer-data.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+} = require('../omeda-data');
+
+/**
+ * A middleware to resync the IdentityX user data with the Omeda customer data. After syncing data,
+ * sets a cookie to prevent resyncs for the configured timeframe.
+ */
+module.exports = ({
+  brandKey,
+  omedaGraphQLClientProp = '$omedaGraphQLClient',
+}) => async (req, res, next) => {
```
Comment (zarathustra323) [inline packages/marko-web-omeda-identity-x/middleware/resync-customer-data.js]: Use the `asyncRoute` util

File: packages/marko-web-omeda-identity-x/add-integration-hooks.js [hunk 3]
```diff
@@ -34,17 +33,6 @@ module.exports = ({
     }),
   });
 
-  idxConfig.addHook({
```
Comment (zarathustra323) [inline packages/marko-web-omeda-identity-x/add-integration-hooks.js:37]: Check if token is set before querying the GraphQL API in `loadActiveContext`

File: packages/marko-web-omeda-identity-x/middleware/resync-customer-data.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  const { identityX, [omedaGraphQLClientProp]: omedaGraphQLClient } = req;
+  if (!omedaGraphQLClient) throw new Error(`Unable to load the Omeda GraphQL API from the request using prop ${omedaGraphQLClientProp}`);
+
+  // Load the user and find their external ID
+  const { user } = await req.identityX.loadActiveContext();
+  const encryptedCustomerId = findEncryptedId({ externalIds: getAsArray(user, 'externalIds'), brandKey });
```
Comment (zarathustra323) [inline packages/marko-web-omeda-identity-x/middleware/resync-customer-data.js]: Add `if(!user || !user.id) return next();` Then remove the user from the if check on L27

File: packages/marko-web-omeda-identity-x/middleware/resync-customer-data.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
```
Comment (zarathustra323) [inline packages/marko-web-omeda-identity-x/middleware/resync-customer-data.js:26]: Add `if (!encryptedCustomerId) return next();`

File: packages/marko-web-omeda-identity-x/middleware/resync-customer-data.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  // Only check for the cookie if we have a valid IdX user and Omeda customer
+  if (user && encryptedCustomerId) {
+    const cookie = resyncCookie.parseFrom(req);
+    // If the cookie doesn't exist (or has expired), resync the user data
+    if (encryptedCustomerId && !cookie) {
```
Comment (zarathustra323) [inline packages/marko-web-omeda-identity-x/middleware/resync-customer-data.js]: Same here, bail early.

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

File: packages/marko-web-identity-x/service.js [hunk 7]
```diff
@@ -32,7 +32,8 @@ class IdentityX {
    * @returns {Promise<object>}
    */
   async loadActiveContext() {
-    if (!this.activeContextQuery) {
+    // Require a token/cookie to check active context. This disables team context from IP/CIDR.
+    if (!this.activeContextQuery && this.token) {
```
Comment (zarathustra323) [inline packages/marko-web-identity-x/service.js]: For clarity, move to a line above: `if (!this.token) return {}`

**Review by zarathustra323 — CHANGES_REQUESTED** [review]
Also, I see there are some `@todo` items listed... should these be addressed now?

**Comment by solocommand** [comment]
@zarathustra323 Done, those are accounted for but left the comments by accident

**Review by zarathustra323 — APPROVED** [review]

**Pull request merged**